Implement Xios handler as a singleton

joewallwork opened the pull request:
Note that this PR merges into the branch for #757, not `develop`.

This PR implements the `Xios` handler object as a singleton, as requested. This means that nextSIM-DG classes don't need to own pointers to the current working `Xios` instance. There can be only one `Xios` instance. The first time the static `Xios::getInstance` method is called, it creates the singleton and returns a pointer to it. On subsequent calls, this method simply returns a pointer to the singleton.

Unfortunately, I had to rework the XIOS read/write test. The singleton design pattern isn't suitable for the way it's currently written, so I re-separated out the reading and writing parts and re-combined the 2D and 3D cases.

joewallwork committed:
Define Xios handler as singleton

joewallwork committed:
Use singleton in most XIOS tests

joewallwork committed:
Use Xios singleton in other tests

joewallwork committed:
Recombine dims; re-separate read and write

joewallwork committed:
Enable XIOS in tests

joewallwork committed:
Revert date; drop unused mutex include

TomMelt reviewed (approved):
Thanks @jwallwork23, looks good to me. I have tested it locally and all tests pass.

timspainNERSC commented on core/src/include/Xios.hpp:
```diff
@@ -30,12 +30,37 @@ namespace Nextsim {
 class Xios : public Configured<Xios> {
-public:
+private:
+    static Xios* instancePtr;
```
Avoid static members in classes. Wrap this in a function to allow it to at least be lazily constructed. See for example [Configurator.hpp:185](https://github.com/nextsimhub/nextsimdg/blob/100e2b742713ce6a809bc2d7fa9386864a1f1852/core/src/include/Configurator.hpp#L185)

Making the pointer a static variable in `getInstance()` would achieve this.

joewallwork replied:
Thanks for this suggestion and for pointing to an example @timspainNERSC. This is a much neater implementation. Done in 10d6207d42813c5009e7a69bba47c07bfd05f110.

timspainNERSC commented on core/src/include/Xios.hpp:
```diff
@@ -30,12 +30,37 @@ namespace Nextsim {
+     *
+     * NOTE: The arguments will only be used the first time this is called.
+     */
+    inline static Xios* getInstance(const std::string dt = "P0-0T01:00:00",
```
Most of the changes in this PR would be unnecessary if references were used instead of pointers. Change the return type to `Xios& getInstance()`.

joewallwork replied:
See 10d6207d42813c5009e7a69bba47c07bfd05f110.

timspainNERSC commented on core/src/include/Xios.hpp:
```diff
@@ -30,12 +30,37 @@ namespace Nextsim {
+        const std::string calendarType = "Gregorian")
+    {
+        if (isNull()) {
+            instancePtr = new Xios(dt, contextId, startTime, calendarType);
```
If the `Xios` class was a static function variable, then it wouldn't have to be dynamically allocated.

joewallwork replied:
Again, see 10d6207d42813c5009e7a69bba47c07bfd05f110.

timspainNERSC reviewed (changes requested):
This needs serious re-working. Most of the changes would be unnecessary if a reference to the `Xios` class was used instead of returning a pointer.

joewallwork committed:
Implement getInstance method as a reference to static Xios instance

joewallwork committed:
Remove added line [skip ci]

joewallwork commented:
[Cherry-picked @TomMelt's NetCDF CI fix from 9d36ed17183e74d8c034a05b3dcf829faafa4469 to see if it fixes the NetCDF/MPI issue this PR gave in the CI just now.]

joewallwork commented:
> [Cherry-picked @TomMelt's NetCDF CI fix from [9d36ed1](https://github.com/nextsimhub/nextsimdg/commit/9d36ed17183e74d8c034a05b3dcf829faafa4469) to see if it fixes the NetCDF/MPI issue this PR gave in the CI just now.]

Hm it seems it does not! will investigate.

joewallwork commented:
> > [Cherry-picked @TomMelt's NetCDF CI fix from [9d36ed1](https://github.com/nextsimhub/nextsimdg/commit/9d36ed17183e74d8c034a05b3dcf829faafa4469) to see if it fixes the NetCDF/MPI issue this PR gave in the CI just now.]
> 
> Hm it seems it does not! will investigate.

Ah, I do get this failure locally. I was running in the wrong build directory...

joewallwork committed:
Avoid setting defaults for constructor twice

joewallwork committed:
Prevent copying

joewallwork committed:
Fix reference usage [skip ci]

joewallwork committed:
Attempt at setting up finalizer

joewallwork committed:
Call finalizer in tests

joewallwork commented:
[Rebased on top of `issue756_xios-calendar-increment`.]

joewallwork committed:
Set up `Finalizer` for XIOS (#822)

Supersedes #808.

As discussed in the meeting this morning, this PR hooks the `Xios` class
into the `Finalizer`. Since no `Model`s are created in the standalone
XIOS unit tests, they need to call `Finalizer::finalize` explicitly. It
seems the same is also true of the `ConfigOutput` and `ParaGrid` tests.
So actually the changeset isn't that different than @TomMelt's one in
#808.

timspainNERSC commented on core/test/ConfigOutput_test.cpp:
I wasn't envisaging using XIOS in ConfigOutput, but instead having a separate class (`XiosOutput`, for example) for handling output when XIOS is being used.

timspainNERSC reviewed (approved):
Approved, with a note on the use of XIOS in `ConfigOutput`, which does not require any action at this point.

joewallwork commented:
> Approved, with a note on the use of XIOS in `ConfigOutput`, which does not require any action at this point.

Great, thanks @timspainNERSC!
